### PR TITLE
PlayChain keychain emulation support

### PR DESCRIPTION
**Pull request by ohaiibuzzle**
This PR propose a new feature that allows apps with PlayTools installed to partially use an Apple Keychain-like APIs to store secrets the same way they usually do. This mitigates the need to relax AMFI for keychain operations

### CODE REVIEW

File: PlayTools/MysticRunes/PlayedApple.swift [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+//        // Get the keychain folder
+//        let keychainFolder = FileManager.default.urls(for: .documentDirectory,
+//                                                          in: .userDomainMask)
+//            .first?.appendingPathComponent("Keychain")
+//
+        let bundleID = Bundle.main.infoDictionary?["CFBundleIdentifier"] as? String ?? ""
+        let keychainFolder = URL(fileURLWithPath: "/Users/\(NSUserName())/Library/Containers/io.playcover.PlayCover")
+            .appendingPathComponent("PlayChain")
+            .appendingPathComponent(bundleID)
+
```
Comment (Depal1) [inline PlayTools/MysticRunes/PlayedApple.swift:28]: Trailing whitespace.

Comment (ohaiibuzzle) [reply PlayTools/MysticRunes/PlayedApple.swift]: Yep, caught it, thanks!

File: PlayTools/MysticRunes/PlayedApple.swift [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        // Create the keychain folder if it doesn't exist
+        if !FileManager.default.fileExists(atPath: keychainFolder.path) {
+            do {
+                try FileManager.default.createDirectory(at: keychainFolder,
+                                                        withIntermediateDirectories: true,
+                                                        attributes: nil)
+            } catch {
+                if PlaySettings.shared.settingsData.playChainDebugging {
+                    NSLog("BUZZINC: Failed to create keychain folder")
+                    }
```
Comment (KhoraLee) [inline PlayTools/MysticRunes/PlayedApple.swift:38]: Trailing curly brackets for debugging is indented wrongly all around the code
Btw how about extracting these debugging code to new method?
```Swift
@objc static public func playChainDebug(_ message: String) {
     if PlaySettings.shared.settingsData.playChainDebugging {
         NSLog(message)
     }
 }
```

Comment (ohaiibuzzle) [reply PlayTools/MysticRunes/PlayedApple.swift]: Done, for both new files

File: PlayTools/MysticRunes/PlayedApple.swift [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+// Implementation for PlayKeychain
+// World's Most Advanced Keychain Replacement Solution:tm:
+// This is a joke, don't take it seriously
+
+public class PlayKeychain: NSObject {
+    static let shared = PlayKeychain()
+
+    private static func getKeychainDirectory() -> URL? {
+        let bundleID = Bundle.main.infoDictionary?["CFBundleIdentifier"] as? String ?? ""
+        let keychainFolder = URL(fileURLWithPath: "/Users/\(NSUserName())/Library/Containers/io.playcover.PlayCover")
```
Comment (JoseMoreville) [inline PlayTools/MysticRunes/PlayedApple.swift:20]:
``` swift
let libraryRoute = FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).first
let keychainFolder = libraryRoute?.appendingPathComponent("Containers/io.playcover.PlayCover/PlayChain").appendingPathComponent("PlayChain").appendingPathComponent(bundleID)
```

This is one more line of code but should reduce hardcode. It's totally just syntax sugar.

Comment (ohaiibuzzle) [reply PlayTools/MysticRunes/PlayedApple.swift]: Done and done!

Comment (ohaiibuzzle) [reply PlayTools/MysticRunes/PlayedApple.swift]: Uh oh. Seems like that will just break the centralized keychain storage. I forgot that the FileManager here will be from the process's perspective, so the one above will actually resolve into `/Users/<User>/Library/Containers/<RunningAppContainer>/Data/Library/Containers/io.playcover.PlayCover/PlayChain/<RunningAppBundleID>/`
Not good.

File: PlayTools.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (IsaacMarovitz) [inline PlayTools.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: `Package.resolved` shouldn't be pushed

File: PlayTools/MysticRunes/PlayedApple.swift (hunk 3 above)
Comment (IsaacMarovitz) [reply PlayTools/MysticRunes/PlayedApple.swift:20]: Change it to `.userDirectory` and change the second line to 
```
libraryRoute?.appendingPathComponent("Library")
                       .appendingPathComponent("Containers")
                       .appendingPathComponent("io.playcover.PlayCover")
```

File: PlayTools/MysticRunes/PlayedApple.swift [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    @objc static public func update(_ query: NSDictionary, attributesToUpdate: NSDictionary) -> OSStatus {
+        // Get the path to the keychain file
+        let keychainPath = keychainPath(query)
+        // Read the dictionary from the keychain file
+        let keychainDict = NSDictionary(contentsOf: keychainPath)
+        debugLogger("Read keychain file from \(keychainPath)")
+        // Reconstruct the dictionary (subscripting won't work as assignment is not allowed)
+        let newKeychainDict = NSMutableDictionary()
+        for (key, value) in keychainDict! {
+            newKeychainDict.setValue(value, forKey: key as! String) // swiftlint:disable:this force_cast
```
Comment (IsaacMarovitz) [inline PlayTools/MysticRunes/PlayedApple.swift:88]: Use an if unwrap rather than force unwrap here

**Review by IsaacMarovitz — CHANGES_REQUESTED** [review]

**Review by IsaacMarovitz — APPROVED** [review]

**Review by Depal1 — APPROVED** [review]

**Pull request merged**